### ISSUE TITLE
Does not appear to work with GM4

### ISSUE THREAD

**Issue by poke** [sizzlemctwizzle/GM_config#97]
I understand that this library likely isn’t actively maintained any more but I just wanted to post this here in case anyone else stumbles on issues when using Greasemonkey 4 (GM4) in Firefox, since this came up [in a different issue](https://github.com/Purfview/IMDb-Scout-Mod/issues/10).

GM_config currently does not work in GM4, likely due to how it activates its user scripts compared to other extensions like Tampermonkey or Violetmonkey. In particular, the following snippet fails in a GM4 user script:

```js
var iframe = document.createElement('iframe');
iframe.src = 'about:blank';
iframe.addEventListener('load', function() {
    console.log('iframe.contentDocument', this.contentDocument); // this is null with GM4
});

document.body.appendChild(iframe);
```

After loading, the frame’s `contentDocument` will be `null` because the same-origin protection will be in place. Basically, it cannot access the document for the iframe that was just created.

Unfortunately, exactly this is done by GM_config in order to load its configuration window:

https://github.com/sizzlemctwizzle/GM_config/blob/a4a49b47ecfb1d8fcd27049cc0e8114d05522a0f/gm_config.js#L325-L330

As a result `GM_config.open()` silently fails, even without logging anything into the browser console.

**Comment by Martii** [comment]
> I understand that this library likely isn’t actively maintained any more

First off never assume. We are in a pandemic.

> GM_config currently does not work in GM4

Half true with the default iframe. Embedded overridden still works with https://openuserjs.org/scripts/Marti/OUJS-1 by going to https://openuserjs.org/user/preferences when signed in *(this is my unit test with a little functionality that I see a portion of usually daily)*... so this is the first I've heard of it not working "out of the box".

Let's see what Anthony says at greasemonkey/greasemonkey#3104 . I'm extremely occupied in real life at the moment so I will attempt to twiddle with this at some point. GM Port latest with SM 2.53.5.1 *(do see and update so will have to recheck... works in 2.53.6 albeit Port is GM3 equivalent)* appears to work with an iframe with @sizzlemctwizzle 's https://openuserjs.org/scripts/sizzle/The_GM_config_Unit_Test .  

I suspect this is a Fx issue since I back-installed older GM4 a few times and retested his unit test. Whether or not Anthony can "shim" in a fix to GM4, or recommend something different, we'll see...  then we'll see what's next here *(have any ideas while I'm AFK? PR perhaps to work around this?)*

**Comment by poke** [comment]
My apologies, I didn’t mean to assume anything 😊 

As for what works and what doesn’t, I honestly don’t know. I have never used GM_config myself and only checked it out because it came up in an (unrelated) issue. Debugging that user script made me dig into GM_config and then I used the advertised “Unit Test” to double check its functionality and figure out why it doesn’t work.

If there are other parts of the library that are unaffected, that’s good. But since I am not a user of the library, I don’t know what else it may or may not do.

**Comment by Martii** [comment]
@arantius

> Are you saying that it's never worked, or recently stopped working?

Apologies for the mention and please feel free to respond on GM proper.

[Sizzles Unit Test](https://openuserjs.org/scripts/sizzle/The_GM_config_Unit_Test) used to work with GM4... however I haven't retested it until yesterday in a very long while. Rolled back to 4.5 yesterday and tried it with no luck. Also have tried your 2021.1.18.15493nightly.

**Comment by Martii** [comment]
I'm also wondering if this could be related to a `window.document` prefix issue similar to the one a few years back. I'll have to dig up that specific patch.

**Comment by Martii** [comment]
First instance where Fx actually notifies on an error:
**67.0.1**
GM 4.9
**FAIL** `TypeError: frame.contentDocument is null ExtensionContent.jsm:993:77`

--- 

Change in Fx behavior:
**58.0.1**
GM 4.0 and 4.10 *(auto-updated)*
**FAIL** TypeError: frame.contentDocument is null

----

Works here:
**57.0.1**
GM 4.0

**PASS**

**Comment by Martii** [comment]
So @ everyone 

We're back to my original assumption with:

> I suspect this is a Fx issue since I back-installed older GM4 a few times and retested his unit test. Whether or not Anthony can "shim" in a fix to GM4, or recommend something different, we'll see... then we'll see what's next here

A portion of it still works if you insert the configuration into a DOM node *(what I call a hook node)* since OUJS-1 is working currently... but OUJS also doesn't use any extra header protection... so not sure how that would pan out for sites that do.

**Comment by arantius** [comment]
https://github.com/greasemonkey/greasemonkey/issues/3104#issuecomment-768691507

**Comment by Martii** [comment]
Thank you Anthony... never would have guessed that on removal of the `src`. Guess the context is inherited from GM extension context with Fx + GM4.

Tested PASS on all current 64 bit browsers of:

* Chromium 88.0.4324.96 (Official Build) snap + TM 4.11   *(should be the same for Chrome)*
* Chromium 88.0.4324.96 (Official Build) snap + VM 2.12.9   *(should be the same for Chrome)*
* Fx 84.0.2 + GM 4.10.0
* Fx 84.0.2 + TM 4.11.6120
* Fx 84.0.2 + VM 2.12.8
* PM 28.17.0 + GM "Fork Branch" 3.31.4
* SM + GM Port 3.10.0.1rc3pre

Not yet tested on:
* Falkon 3.1.0 *(Qt 5.12.8)* + native GreaseMonkey *(this will take a bit since I haven't been in that for a long time and don't recall if `@require` is supported)*
* Safari ??? @derjanb Would you please test https://openuserjs.org/scripts/Martii/The_GM_config_Unit_Test installation and go to https://openuserjs.org/scripts/sizzle/The_GM_config_Unit_Test and see if the settings dialog pops up. I haven't worked through my Apple ID *(took them 13 months to delete my old one heh)* yet and that machine is almost always off so I don't have TM yet for dev purposes... and let me know. I would really appreciate it and thanks so much in advance. :)
* Direct DOM test for website hosting
* Windows *(VM time after Qt is finished downloading... whistles)*
* Need to try `@grant none` as well in XUL based user.js engines and GM4 *(unless the sandboxing has changed from before with needing to do `unsafeWindow` tests)*

Will update this comment on what I do get done testing with.

**Comment by Purfview** [comment]
Win7 + The GM_config Unit Test 2.2.2.3a:

Those below PASS:

Firefox 86.0b2 + GM 4.11
Firefox 86.0b2 + VM 2.12.8
Firefox 86.0b2 + TM 4.11.6120
Firefox 86.0b2 + TMb 4.12.6126

Waterfox Classic 2021.01.1 + GM 3.17
Waterfox Classic 2021.01.1 + VM 2.12.8

Couldn't continue testing more as I get this error now (same after an hour):
429 - Too Many Requests at https://openuserjs.org/src/libs/martii/GM_config.min.js

**Comment by Purfview** [comment]
After 6 hours my IP is still banned on openuserjs.org, how long it lasts?

Here  some test from another computer.

**PASS:**
Waterfox Classic 2021.01.1 + GM 4.11
Pale Moon 28.17.0 + GM 3.31.4

**FAIL:**
Opera 73.0.3856.344 + TM 4.11.6120  (GM_config Unit Test 2.2.2.2 works)
Opera 73.0.3856.344 + VM 2.12.9  (GM_config Unit Test 2.2.2.2 works)

**Comment by Martii** [comment]
@Purfview 

> After 6 hours my IP is still banned on openuserjs.org, how long it last?

It's supposed to be [~60 minutes](https://github.com/OpenUserJS/OpenUserJS.org/blob/c6b96c4464681a484753986b1cb2eeac8bc6508f/routes.js#L35) ... as long as you don't hit it too much beyond the notice... I'll have to see if that dep is messing up or needs loosening. I just dropped the database locks there so that should reenable it for another period set of what it's currently doing. i.e. reenabled for install on all ip's. Apologies. Some user.js engines might be hitting it twice per install again too. \*Adds to todo check list again\*

Waterfox is what I was concerned with as I don't have that browser. Good to have Win7 though. Thank you. 

I'm still waiting on macOS to update so I can test Fx/Chrome/Opera there too... and I won't be able to test Safari until I get together with @derjanb . My apple ID is still foobarred on their end plus the 5 hour Big Sur *(11.1)* update in-between  winks, errands, and work irl.

I'm leaning towards Opera just being too old or bugged since the vast majority of platforms/agents are passing. So a PR could be in the works. I'll look into a console message too but that can get sticky with logging compatibility.

Thanks again. :)

**Comment by derjanb** [comment]
> Safari ???

@Martii will check this soon... Sorry.

**Comment by Martii** [comment]
@derjanb 

No hurries... I appreciate the visit. :) Not willing to spend $5000.00 on a new mac just yet *(or ever I think ;)*

**Comment by Purfview** [comment]
Initially I hit beyond "the notice", which I didn't see because I was just getting a blank page, "429 - Too Many Requests" I saw only when I tried to download the script directly with browser, "Pale Moon 28.17.0 + GM 3.31.4" gave me some message like "Try after "3660". Now my IP is unbanned.

Opera is based on Chromium and is updated regularly,  imho Opera support is more important than GM4 support as Fx has other working monkeys.
I found this page for bug reports:
https://help.opera.com/en/computer-bug-wizard

**Comment by Martii** [comment]
> Opera support is more important than...

Well this is going to take some time. Lots of failures so far in macOS *(more tests to go)*. I'll see if I can silently test out a version that recreates the iframe without source on `null` however one might be able to do a custom DOM node like OUJS-1 just make it an iframe... never tried that before so it might work.

**Comment by derjanb** [comment]
@Martii No dialog at Safari 14. @arantius' Test script "iframe contentDocument" works at Safari as well.

![image](https://user-images.githubusercontent.com/767504/106210859-f020ce00-61c7-11eb-8a4b-9b3e8520f561.png)

**Comment by Martii** [comment]
@derjanb 

Thank you. If you have a chance one more please... Make sure GM_config Unit Test is uninstalled/disabled first; Install at https://openuserjs.org/scripts/Martii/The_GM_config_Unit_Test2 ; Then revisit sizzles script homepage at https://openuserjs.org/scripts/sizzle/The_GM_config_Unit_Test... this is my proposed fix. 

Cc: @Purfview  ... you too... test only what you think is necessary. This makes everything under Linux pass and I even tested Edge and it passes too which was unexpected... but the patch is slightly different than what Anthony mentioned. I redropped the install limiter database so theoretically you should have some more freebies... will have to look into why that's being overly aggressive. Please feel free to try [it](https://openuserjs.org/libs/Martii/GM_config2) with your [main script linkage](https://github.com/Purfview/IMDb-Scout-Mod/issues/10) temporarily too but note it will be going away when this issue is closed since I'll merge it here. Thanks.

Still have to test it direct with a site hosting but I suspect since it's so minimal of a change it should still work.... and I should try it with an open/close/open/close button to be ultra sure.

**Comment by derjanb** [comment]
> Install at https://openuserjs.org/scripts/Martii/The_GM_config_Unit_Test2 ; [...] this is my proposed fix.

@Martii This worked. 👍

**Comment by Purfview** [comment]
@Martii
The_GM_config_Unit_Test2 works on all browsers I've tested above and IMDb Scout Mod works too.
Btw it fixed another issue: GM_config window was borked because it was loading inside iframes' of the ads, I had it fixed just by removing the ads from IMDb page.

===================================================

I've noticed unrelated IMDb Scout Mod issue:  Page doesn't reload after Close button is pressed (on Chrome and Opera). Maybe  you have an advice on that? Current code looks like this:

`document.getElementById('imdb_scout').contentWindow.document.getElementById('imdb_scout_closeBtn').onclick = function () {window.location.reload(false)};`

(Note: "imdb_scout" is id of GM_config)

**Comment by Martii** [comment]
@derjanb 
Thank you so much. I think I'm going to have to try resetting my mac to clear the Apple ID. Ughh. I really, really, really appreciate your help.

@Purfview 
Awesome.... I'll try to get this merged in a bit after one more test before I exit for the weekend.

> Maybe you have an advice on that? ...

Have you tried putting your reload in the `onClose` callback instead?

**Comment by Martii** [comment]
Merged!

http://sizzlemctwizzle.github.io/GM_config/ appears to be working... glad @sizzlemctwizzle had his webhook enabled for the site hosting test at http://sizzlemctwizzle.github.io/GM_config/ *([Secure](https://sizzlemctwizzle.github.io/GM_config/))*

I'll keep the issue open for a little bit in case there is some unknown issue that pops up. Works with OUJS-1 as well as it should since it doesn't use the default `iframe`. I'm overly cautious with changes since I'm the trailing dev here.

And a final thank you for @poke for being clever enough to find the minimal steps reproducible and opening the issue. :)

**Comment by Purfview** [comment]
> 
> 
> Have you tried putting your reload in the [`onClose`](https://github.com/sizzlemctwizzle/GM_config/wiki/Event-Callbacks#onclose) callback instead?

That worked, somehow I missed it.  Thanks for the hint.

**Comment by Purfview** [comment]
Looks like I'm banned again on openuserjs, I wasn't doing anything unusual.
If you need IP then look at user: SpiderMunkey.

**Comment by Martii** [comment]
@Purfview 

> Looks like I'm banned again on openuserjs

Okay relooked at what was done... 50 times per hour per install/raw... so two scripts *(.user.js and .js)* means 25 times per hour... if the user.js engine is bugged by calling twice then it's 12.5 *(rounded down because it doesn't do decimals... so 12)*. Once every 5 *( 60 / 12 )* minutes or so is about the maximum you can tweak your script with a bugged .user.js engine... or 11 requires. That's more than enough but would be better if any .user.js engines aren't bugged atm. Do more limited testing over a longer period *(hence why I said earlier about "test only what you think is necessary")*. This is by design... does make testing a brute aweful slow but you should see GH's rate limiting. ;)

Also if you are using Chromium products they tend to always retry on failure status... which isn't a good thing imho. As a test for you... try a url without your wifi/ethernet plugged in and then plug in/turn it on... and watch Chromium products auto-refresh. Booo!

> If you need IP...

Doesn't work that way... it's obfuscated in the DB from last check.

I haven't seen any further issues so I'm going to ahead and close this. If the main issue subject re-pops up open a new issue please referencing this one. Also retested on Fx for Android 68 something *(the last one that supported extensions)* and it's happy so far.
